Framework/MDAlgorithms/src/CompareMDWorkspaces.cpp

```diff
@@ -31,6 +32,182 @@ class CompareFailsException : public std::runtime_error {
 namespace Mantid {
 namespace MDAlgorithms {
 
+class SimpleMDEvent {
+
+private:
+  std::vector<float> mCoordinates;
+  float mSignal;
+  float mError;
+
+public:
+  SimpleMDEvent(const std::vector<float> &coordinates, const float &signal, const float &error)
+      : mCoordinates(coordinates), mSignal(signal), mError(error) {
+    //        mCoordinates = coordinates;
+    //        mSignal = signal;
+    //        mError = error;
+  }
+
+  SimpleMDEvent(const SimpleMDEvent &other)
+      : mCoordinates(other.mCoordinates), mSignal(other.mSignal), mError(other.mError) {
+    // mCoordinatesmCoordinates = other.mCoordinates;
+    // mSignal = other.mSignal;
+    // mError = other.mError;
+  }
+
+  // pretty output
+  std::string str() const {
+    std::stringstream ss;
+    size_t numdirs = mCoordinates.size();
+    for (size_t i = 0; i < numdirs; ++i)
+      ss << mCoordinates[i] << ", ";
+    ss << "signal = " << mSignal << ", error = " << mError;
+
+    return ss.str();
+  }
+
+  float getSignal() const { return mSignal; }
+
+  float getError() const { return mError; }
+
+  bool operator()(SimpleMDEvent &event1, const SimpleMDEvent &event2) {
+    // std::cout << "() is called\n";
+    return event1 < event2;
+  }
+
+  /**
+   * @brief override operator <
+   * order
+   * 1. coordinate
+   * 2. signal
+   * 3. error
+   * @return
+   */
+  bool operator<(const SimpleMDEvent &event2) const {
+
+    bool less(true);
+    bool equal(false);
+
+    // compare coordinates
+    size_t numdirs = mCoordinates.size();
+    for (size_t i = 0; i < numdirs; ++i) {
+      // equal is hard to tell
+      coord_t diff = mCoordinates[i] - event2.mCoordinates[i];
+
+      if (fabs(mCoordinates[i]) > 1E-7 && fabs(diff / mCoordinates[i]) < 1E-6) {
```

Comment:
If the box structure of the two workspaces to compare must be the same, then I would to the event comparison within each leaf box since these boxes have the smallest event population.

Iterate over the leaf boxes, and for each leaf box of MDEventWorkspaces "A" and "B" I would iterate over the event dimensions, doing:
- pick the coordinates of the selected dimension of each event and sort them (smallest to biggest), yielding ordered lists `La` and `Lb` of scalars. The lists should be the same length, and then `abs(La[i] - Lb[i]) < EPSILON` for all `i` or just `sum(abs(La[i] - Lb[i])) < N * EPSILON`, where `N` is the length of the lists.
- If the previous comparison return `true` then do the same with the coordinates of the next dimension.
If the event comparison requires comparing other properties than the dimension coordinates (e.g. `signal` and `error`), the same scheme can be applied.

The previous comparison scheme is not exactly the same as comparing each event of workspace `A` to each event of workspace `B` but when one has millions of uncorrelated events the previous comparison scheme will fail is the set of events is not the same for `A` and `B`

Reply:
The implementation of this `ugly` code is that in system test, some `gold` MDEvent does not have adequate precision.